[블랙잭 1단계] 메다 미션 제출합니다.

medAndro opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [ ] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

안녕하세요 둘리!

- 이번에 미션을 진행하면서 페어와 함께 TDD로 구현하여 정상적으로 동작시키는것을 최우선 목표로 삼았고 아직은 미숙하지만 모든 과정에서 최대한 TDD로 구현하도록 노력했습니다 현재의 TDD로 구현하고 있는 방식이 둘리가 보시기에 어떻게 생각하는지 궁금합니다.

- 현재로서는 각각의 객체들이 너무 많은 책임을 가지고 있고 패키지도 올바르게 분리하지 않았다고 생각합니다. 개선을 시키기 위한 인사이트를 던져주시면 감사하겠습니다!

- 가장 마지막에 구현한  `WinLossStatistics` 클래스의 로직들은 TDD로는 구현하였으나 상수를 하드코딩하고 if문이 난무하고 메서드의 라인수가 길어지는 문제, `GameParticipant`의 `abstract fun play()`의 미사용, 예외 처리의 부재 등 여러 시간부족 이슈등으로 발생한 문제들을 인지하고 있으며 차차 개선해 나가고자 합니다!

medAndro committed:
feat: 코틀린 DSL 실습 코드 추가

medAndro committed:
docs: card와 player 기본 기능 요구사항 작성

medAndro committed:
test: Card 생성 테스트 코드 구현

medAndro committed:
feat: 카드 클래스의 유효성 검증 로직 작성

medAndro committed:
chore: .gitkeep 삭제

ijh1298 committed:
docs: Deck의 무작위 카드를 뽑는 메서드 요구사항 추가

ijh1298 committed:
test: Deck 무작위 카드 뽑기 테스트 코드 구현

ijh1298 committed:
docs: Card를 index로 생성하는 요구사항 추가

ijh1298 committed:
test: Card를 index로 생성하는 테스트 코드 구현

ijh1298 committed:
feat: Card index를 인자로 받는 생성자 추가

ijh1298 committed:
docs: Card를 index로 생성하는 요구사항 완료 체크

ijh1298 committed:
test: Deck에서 무작위 카드를 뽑는 테스트 코드 구현

ijh1298 committed:
refactor: 테스트 명과 기능 요구사항 수정

ijh1298 committed:
feat: Deck 클래스 기본 구조 구현

medAndro committed:
docs: Deck에서 카드를 뽑는 요구사항 완료 체크

medAndro committed:
chore: todo 추가

medAndro committed:
test: 플레이어가 카드를 2장 받는 테스트 코드 구현

medAndro committed:
refactor: Deck을 전역적으로 사용하도록 수정

medAndro committed:
docs: player가 보유한 카드를 보여주는 요구사항 추가

medAndro committed:
refactor: 플레이어가 카드를 가져오는 테스트내의 메서드 수정

medAndro committed:
feat: Player 클래스 기본 구조 구현

medAndro committed:
docs: 카드번호, 문양, 룰에 관한 요구사항 추가

medAndro committed:
test: 카드번호의 문자열 반환 테스트 코드 구현

medAndro committed:
feat: CardNumber 클래스 기본 구조 구현

medAndro committed:
test: 문양의 문자열 반환 테스트 코드 구현

medAndro committed:
feat: Suit 클래스 기본 구조 구현

medAndro committed:
refactor: Card의 유효성 검증 로직을 Suit, CardNumber 로 이동

ijh1298 committed:
docs: 카드 번호, 문양을 받아오는 요구사항 완료 체크

ijh1298 committed:
refactor: domain 패키지 분리

ijh1298 committed:
docs: 손패의 결과를 물어보는 테스트 코드 구현

ijh1298 committed:
test: 손패의 결과를 물어보는 테스트 코드 수정

ijh1298 committed:
refactor: CardNumber의 가시성 변경자 수정

ijh1298 committed:
feat: Rule에 손패 결과 값을 구하는 메서드 구현

ijh1298 committed:
docs: Rule에 손패 결과 값을 구하는 요구사항 완료 체크

medAndro committed:
chore: 요구사항 명명 수정

medAndro committed:
test: Rule에게 플레이어의 드로우 여부를 받아오는 테스트 코드 구현

medAndro committed:
feat: Rule에서 플레이어의 드로우 가능 여부를 반환하는 메서드 구현

medAndro committed:
docs: Rule에서 플레이어의 드로우 가능 여부를 반환하는 요구사항 완료 체크

medAndro committed:
test: Rule에게 딜러의 드로우 여부를 받아오는 테스트 코드 구현

medAndro committed:
feat: 룰에게 딜러의 드로우 여부를 받아오는 메서드 구현

medAndro committed:
docs: 룰에게 딜러의 드로우 여부를 받아오는 요구사항 완료 체크

ijh1298 committed:
refactor: Deck에서 카드를 주는 메서드명 변경

ijh1298 committed:
test: Deck 메서드명 수정에 따른 테스트 코드 변경

ijh1298 committed:
feat: 플레이어와 딜러의 공통 역할의 GameParticipant 추상 클래스로 구현

ijh1298 committed:
refactor: Rule의 플레이어의 드로우 여부를 받는 메서드명 변경

ijh1298 committed:
test: 플레이어가 카드를 뽑는 테스트 코드 구현

ijh1298 committed:
refactor: GameParticipant의 필드 가시성 변경자 수정

ijh1298 committed:
feat: Player가 GameParticipant를 받도록 구현

ijh1298 committed:
docs: 플레이어가 덱에서 카드를 뽑는 요구사항 완료 체크

ijh1298 committed:
refactor: GameParticipant 역할 재정의

medAndro committed:
feat: Input View 구현

medAndro committed:
feat: OutputView 일부 구현

medAndro committed:
docs: Card가 자신의 정보를 문자열로 주는 기능 요구사항 추가

medAndro committed:
test: 카드 정보를 문자열로 만드는 테스트 코드 구현

medAndro committed:
feat: 카드 정보를 문자열로 반환하는 메서드 추가

medAndro committed:
feat: OutputView의 손패의 카드 정보를 출력하는 메서드 구현

medAndro committed:
feat: 컨트롤러의 Player 초기화 단계 구현

medAndro committed:
feat: 딜러의 기본 구현 및 isDrawFinish 추상화 메서드를 쓰는 모든 코드 구현

medAndro committed:
refactor: 카드 배부 메시지의 자료형 수정

medAndro committed:
refactor: 딜러의 카드 정보를 한장만 출력하도록 로직 수정

ijh1298 committed:
fix: Rule의 isBurst 메서드 판단 조건 수정

ijh1298 committed:
feat: 카드의 실제 값을 반환하는 메서드 추가

ijh1298 committed:
refactor: 카드 점수 계산에 실제 값을 반환하는 메서드를 호출

ijh1298 committed:
feat: OutputView 딜러의 드로우 메세지 출력 메서드 추가

ijh1298 committed:
feat: Casino에 플레이어와 딜러의 DrawPhase 구현

ijh1298 committed:
feat: OutputView에 모든 참가자들의 카드 결과를 보여주는 메서드 추가

ijh1298 committed:
feat: Casino에 모든 참가자들의 카드 결과를 보여주는 메서드 호출

medAndro committed:
feat: 카드 분배 후 히트하지  않을 경우 카드 정보를 출력하는 코드 추가

medAndro committed:
docs: 카드 정보 출력 요구사항 완료 체크 및 통계의 손패 비교 요구사항 추가

medAndro committed:
test: 손패 두개 비교 테스트 코드 구현

medAndro committed:
chore: 오타 수정

medAndro committed:
feat: WinLossStatics 기본 구조 구현

medAndro committed:
docs: 손패 두개 비교 요구사항 완료 체크 및 딜러 결과 텍스트 받아오는 요구사항 추가

medAndro committed:
test: 딜러의 승 무 패 결과 결과를 텍스트로 반환하는 테스트 코드 구현

medAndro committed:
feat: 최종 승패 로직 구현

medAndro committed:
test: 딜러와 플레이어의 버스트 교착 상태의 테스트 코드 추가

medAndro committed:
fix: 플레이어와 딜러의 승패 계산 로직 정상화

medAndro committed:
docs: 딜러의 전체 승 무패 결과 텍스트를 받아오는 요구사항 체크

medAndro committed:
feat: DSL 연습을 파일 한개로 임시 구현 완료

hyemdooly commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,32 @@
+package blackjack.view
+
+class InputView {
+    fun readPlayerNames(): List<String> {
```
`InputView`에서  `Player`를 생성해서 넘겨주지 않은 이유가 있으신가요?

medAndro replied:
이전 미션에서 `InputView`의 역할은 사용자의 입력을 받아 입력값을 가능한 원형을 유지한 채 컨트롤러로 전송하는것으로  이해했습니다.  
`Player`를 생성하는건 단순히 데이터를 담아서 넘겨주는게 아닌가? 싶기도 했는데 뷰에서 도메인 객체를 만들 때 예외가 발생한다면 도메인의 예외처리를 inputView에서 하게 되어 도메인의 역할을 침범하게 될 것 같다고 생각했습니다.

hyemdooly replied:
아하! InputView에서 도메인을 몰라야한다면, OutputView에서 동일하게 몰라야할 것 같은데 OutputView에서는 참조를 하고 있어서 리뷰 드렸어요.
그렇다고 해서 뷰를 위한 모델을 만드는 것은 지금 단계에서 과할 수도 있다고 생각하는데, 이 부분은 크루들과 같이 얘기해보면 좋을 것 같아요!

medAndro replied:
처음엔 [3기 제리의 테코톡](https://youtu.be/ogaXW6KPc8I?t=236) 을 참고해서 View에서는 모델을 참조하거나 만들어도 되겠구나 라고 생각했었는데 [로또 리뷰 과정](https://github.com/woowacourse/kotlin-lotto/pull/127/files/00ec1f0f347e9f68fe406cf135068d75e602a6d1#r1965102144)에서 참조만 가능하고 생성은 불가능하게 하는게 좋겠다는 결론을 내려서 현재와 같아졌습니다.
다음 단계로 나아가면 DTO를 써야 할 것 같네요 현재단계에선 언급해주신대로 과한것 같구요...!

hyemdooly commented on src/main/kotlin/blackjack/controller/Casino.kt:
```diff
@@ -0,0 +1,71 @@
+import blackjack.view.OutputView
+
+class Casino(
+    val inputView: InputView,
+    val outputView: OutputView,
```
inputView, outputView는 외부에서 사용하지 않도록 private 처리를 하면 좋겠어요!

medAndro replied:
이전 로또 미션에서도 받은 리뷰라 반성중입니다...

hyemdooly commented on src/main/kotlin/blackjack/domain/model/CardNumber.kt:
```diff
@@ -0,0 +1,34 @@
+package blackjack.domain.model
+
+@JvmInline
+value class CardNumber(
```
suit, cardNumber를 상수화한다면 어떨까요?
스페이드, 하트, 다이아몬드, 클로버 / ACE, J, Q, K,... 더 명시적으로 표현할 수 있고, 점수도 갖고 있게 할 수 있으니, 점수 계산할 때도 편할 것 같아요!

medAndro replied:
...를 하고 싶었는데 시간 부족으로 미뤄둔 부분이라 열심히 반영해보겠습니다

hyemdooly commented on src/main/kotlin/blackjack/domain/model/Rule.kt:
```diff
@@ -0,0 +1,33 @@
+package blackjack.domain.model
+
+class Rule {
```
class에 companion object만 있다면 class를 object로 만드는 것이 좋겠어요.
추가로, Rule이 하는 일을 각 객체에게 주어, 객체가 스스로 일을 하도록 하면 어떨까요?

medAndro replied:
object라는게 있다는걸 푸시하고 페어와 함께 서로 알게 되었습니다. 
사실 프리코스때도 한번 썼던거같은데 잊어버렸나봐요 😅
PR에서 언급했듯 각각의 객체, 특히 Rule이 너무 많은 일을 하는 것 같아서 책임을 분리하는 방향으로 리팩터링 해봐야 겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/domain/model/Player.kt:
```diff
@@ -0,0 +1,15 @@
+    name: String = DEFAULT_NAME,
+) : GameParticipant(name = name) {
+    override fun play() {
+        // todo
```
todo 주석 활용 👍

medAndro replied:
todo였지만 play는 삭제해야 할것같아요😂

hyemdooly commented on src/main/kotlin/blackjack/domain/model/Deck.kt:
```diff
@@ -0,0 +1,12 @@
+
+class Deck {
+    companion object {
+        private val cards = ArrayDeque(Card.CARD_INDEX_RANGE.shuffled().map { Card(it) })
```
여러 자료구조 중에 `ArrayDeque`를 사용하신 이유가 무엇인지 궁금합니다!

medAndro replied:
카드를 한장씩 Pop해야 하는 상황에서 어떤 자료구조를 쓰면 좋을까 페어와 논의하다가 

> 자료구조적으로는 카드를 담아두고 빼내기 위해서 스택이나 큐처럼 동작하는 자료구조를 사용해야겠다고 생각했습니다. 정확한 차이점은 기억이 안나지만 알고리즘을 풀 때 있어서 자바에서 Stack, Queue 구현체에 단점이 존재한다고 들었었고, 링크드리스트보다는 시간복잡도 면에서 나은 ArrayDeque을 무의식적으로 사용했습니다.
https://github.com/woowacourse/kotlin-blackjack/pull/128#discussion_r1986069929 

라는 페어의 제안으로 사용해보게 되었습니다. 

- ArrayDeque: 양쪽 끝에서 O(1) 시간 복잡도로 효율적인 추가/제거 작업 가능하며 메모리 효율적이지만, 제거 작업을 해도 내부 배열의 크기가 줄어들지 않음
- MutableList: 끝에서 요소 추가/제거시 효율적이지만, 중간이나 처음에서 제거하는 작업은 O(n) 시간 복잡도로 비효율적
- Stack: LIFO(Last-In-First-Out) 원칙을 따르며 한쪽 끝에서만 O(1) 시간 복잡도로 요소 추가/제거가 가능하지만 Java의 레거시 클래스로 성능 이슈 존재
- Queue: FIFO(First-In-First-Out) 원칙을 따르며 한쪽 끝에서 추가하고 반대쪽 끝에서 제거하는 구조로, 자바의 Queue자체는 인터페이스이며 LinkedList나 ArrayDeque로 구현 가능

라는걸 공부하게 되었고 현재로서는 ArrayDeque를 유지할 생각입니다.

hyemdooly commented on src/main/kotlin/blackjack/domain/model/GameParticipant.kt:
```diff
@@ -0,0 +1,21 @@
+        repeat(2) { drawCard() }
+    }
+
+    fun showCards(): List<Card> = cards.toList()
```
`show`이지만 카드를 리턴하는 역할을 하고 있어요. 바꿔볼 수 있을까요?

medAndro replied:
확실히 잘못된 이름이네요 고민해서 바꿔보겠습니다

hyemdooly commented on src/main/kotlin/blackjack/domain/model/WinLossStatistics.kt:
```diff
@@ -0,0 +1,53 @@
+        return WinLoss.WIN
+    }
+
+    fun getDealerWinLossText(): String {
```
아직 이쪽을 더 수정하겠다고 하셨지만 코멘트 조금 남겨봅니다!
출력을 위한 `String` 생성은 도메인이 아닌 뷰의 역할 아닐까요? 🤔

medAndro replied:
확실하게 뷰의 역할이 맞다고 생각합니다
승패 결과 로직은 시간부족으로 일단 돌아가게끔 하는데 집중해 보았습니다! 
OutputVIew에서 처리 시켜 보겠습니다.

hyemdooly commented on src/main/kotlin/blackjack/domain/model/WinLossStatistics.kt:
```diff
@@ -0,0 +1,53 @@
+class WinLossStatistics {
+    private val dealerResults = mutableMapOf(WinLoss.WIN to 0, WinLoss.LOSE to 0, WinLoss.DRAW to 0)
+
+    fun calculatePlayerWinLoss(
```
> 처음 2장의 카드가 에이스와 10(J·Q·K를 포함)으로 21점이 된 것을 '블랙잭'이라고 하며, 베팅액의 1.5배를 얻는다. 딜러가 블랙잭이라면 같은 블랙잭을 가진 플레이어를 제외한 모든 플레이어는 즉시 패배하며 베팅액을 잃는다. 딜러와 플레이어가 동시에 블랙잭인 경우에는 푸시(Push)라 하여 무승부가 된다. 푸시가 될 경우, 해당 플레이어는 자신이 베팅한 금액을 돌려받는다.

딜러가 블랙잭인 경우에도 고려해주셔야 합니다!
딜러가 블랙잭(A + 10인 카드)이고 플레이어가 블랙잭이 아니지만 합이 21인 경우에도 딜러 승리입니다. 🥲

medAndro replied:
이런 룰을 몰랐습니다... 반영 해보겠습니다!

hyemdooly reviewed (changes requested):
안녕하세요, 메다! 이번 미션을 함께 하게 된 둘리입니다. 반갑습니다! 😄 
시간이 부족했지만 어느정도 요구사항을 만족해주셨군요! 💯 
하지만 메다 스스로도 아직 아쉬운 부분이 있는 것 같아, 이번 리뷰는 조금 가볍게 남겼습니다. 메다가 생각하고 있는 개선점까지 포함해서 다음 리뷰 요청을 부탁드릴게요. 🙏 
더 궁금한 점이 있으시면 코멘트 달아주시거나 DM 주세요! 💪

hyemdooly commented:
> 이번에 미션을 진행하면서 페어와 함께 TDD로 구현하여 정상적으로 동작시키는것을 최우선 목표로 삼았고 아직은 미숙하지만 모든 과정에서 최대한 TDD로 구현하도록 노력했습니다 현재의 TDD로 구현하고 있는 방식이 둘리가 보시기에 어떻게 생각하는지 궁금합니다.

테스트 코드를 작성하고, 구현을 한 모습이 보이기는 하는데요! TDD의 모든 단계를 커밋하는 것은 아니니 제가 판단하기에는 쉽지 않긴 합니다. 🥲
TDD를 잘했다면 모든 도메인 로직에 테스트 코드가 잘 되어있을텐데요! 전체적으로 있어야할 테스트 코드는 다 있는 것 같아서 잘하고 계신 것으로 보여요. 😄

medAndro committed:
refactor: 도메인 패키지 역할별 구조화

medAndro committed:
docs: Step1 리팩터링 요구사항 초안 추가

medAndro committed:
refactor: CardNumber를 enum class Number로 수정

medAndro committed:
refactor: Suit를 enum class Suit로 수정

medAndro committed:
docs: HandCards 활용 요구사항 추가

medAndro committed:
feat: 손패(HandCards) 일급컬렉션과 카드 상태(CardStatus) 추가하여 적용

medAndro committed:
docs: 완료 기능 목록 체크 및 추가

- 카드 문양, 번호, 상태 추가
- 기존 코드의 enum 클래스 적용 요구사항 추가
- 카드의 이름 생성 책임을 OutptuVIew로 이동하는 요구사항 추가

medAndro committed:
refactor: 카드 문자열 생성 책임을 OutView 이동

medAndro committed:
refactor: Rule의 역할을 각 클래스로 분리

medAndro committed:
refactor: 딜러의 첫 카드를 가져오는 메서드 추가

medAndro committed:
docs: 리팩터링 목록 추가 및 체크

- 카드의 한국어 이름 텍스트를 만드는 책임을 OutputView로 이동하는 요구사항 체크
- 블랙잭시 딜러 판정과 유효성 감사 요구사항 추가

medAndro committed:
refactor: 불필요한 play 메서드와 승,무,패 텍스트 제거

medAndro committed:
refactor: 카드의 초기화 책임을 HandCards에서 GameParticipant로 이동

medAndro committed:
feat: 승, 무, 패 한국어 텍스트를 받아오는 메서드 추가

medAndro committed:
refactor: 승 무 패 판별 로직 재작성

medAndro committed:
docs: 게임 참여자의 카드를 초기화하는 책임을 컨트롤러로 이동하는 요구사항 추가

medAndro committed:
refactor: 게임 참여자의 카드 초기화 로직을 컨트롤러로 이동

medAndro committed:
refactor: 승 패 통계 테스트코드에 새로운 로직 적용

- 카드 리스트 대신 플레이어와 딜러를 직접 받아서 검사

medAndro committed:
refactor: 딜러의 전체 승 무 패 결과 텍스트 출력 로직 변경

- 승, 무 패 텍스트 생성을 출력 뷰에서 처리하도록 수정

medAndro committed:
refactor: WinLossStatistics의 레거시 코드 삭제

medAndro committed:
test: 승 패 통계의 레거시 테스트 코드 수정 및 추가

medAndro committed:
fix: 플레이어가 블랙잭이고 딜러가 블랙잭이 아닐 때 승리하지 못하는 버그 수정

medAndro committed:
fix: 변경된 요구사항으로 인한 테스트 값 수정

- 컨트롤러에서 직접 2장을 뽑아 초기화하므로 3에서 1로 변경

medAndro committed:
refctor: 기능을 이동하여 필요 없어진 Rule클래스 제거

medAndro committed:
docs: 완료된 요구사항 체크 및 새로운 요구사항 추가

medAndro committed:
﻿docs: todo 제거 요구사항 추가

medAndro committed:
refactor: Casino의 중복 If문 제거

medAndro committed:
feat: 카드 소진시 예외 추가

medAndro committed:
docs: todo 처리 기능목록 완료 체크 및 새로운 기능목록 추가

medAndro committed:
refactor: Deck을 object로 변경

medAndro committed:
feat: 실행 결과와 같도록 라인을 분리하는 코드 추가

medAndro committed:
docs: 완료한 요구사항 체크

medAndro committed:
refactor: 로직에 영향이 없는 자잘한 코드 수정

- 블랙잭 실행 함수명 변경
- 누락된 private 접근 제어자 및 const 추가
- 사용하지 않는 메서드 제거
- 코드내 문자열 상수화

medAndro committed:
docs: 완료된 요구사항 체크

medAndro committed:
refactor: DSL 실습 코드 분리

medAndro committed:
fix: 잘못된 gitIgnore 수정

medAndro committed:
docs: DSL 실습 코드 분리 완료 체크

medAndro committed:
test: HandCard의 테스트 코드 추가

medAndro commented:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

둘리! 안녕하세요

step1 첫 PR이 시간이 부족해서 미흡한점이 많앗고 이전 미션과 병행하다 보니 수정하는데 상당히 오래 걸렸던 것 같습니다. 현재 제가 생각하는 고쳐야 할 부분들은 어느정도 고쳐졌다고 생각해서 다시 요청드립니다. 

enum class인 Number에 ACE부터 KING까지 담아두고 카드의 이니셜은 변하지 않는 값이라고 생각하여 포함시켜 두었는데 관심사의 분리에 따라 출력 뷰로 가야 하나 말아야 하나 고민이 되었습니다. 현재로서는 출력 외에는 사용하고 있지 않은 것 같아요

각 게임 참여자가 가지는 HandCards라는 손패(손에 쥔 패) 일급 컬렉션을 도입하여 각 참여자들의 손패를 관리하도록 했습니다.  가령 WinLossStatistics의 [이 부분](https://github.com/medAndro/kotlin-blackjack/blob/f56ee63a3258ec3843af1e9b1d4aea049f6b6d03/src/main/kotlin/blackjack/domain/model/progress/WinLossStatistics.kt#L16) 과 같이 게임 참여자 인스턴스의 handCards를 불러와 직접 조작시키는 과정이 빈번하게 발생하는데 괜찮은건지 작성하면서 의문이 들었습니다.

마지막으로 코틀린 DSL 어렵네요... 일단 dslstudy 패키지에 만들어 놓긴 했는데 코틀린 DSL 실습이 이걸 의도한건지 모르겠습니다...

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/Deck.kt:
```diff
@@ -0,0 +1,12 @@
+package blackjack.domain.model.card
+
+object Deck {
```
Deck이 object인 이유는 무엇인가요?

medAndro replied:
블랙잭 모듈 내에서 랜덤 카드 인스턴스 한벌로 구성 된 덱을 싱글턴으로 가지고 있으면 좋을 것 같아 object로 사용하였습니다.

hyemdooly commented on src/main/kotlin/blackjack/domain/model/participant/GameParticipant.kt:
```diff
@@ -0,0 +1,18 @@
+
+    fun cardSize() = handCards.currentCards().size
+
+    fun drawCard() {
+        handCards.addCard(Deck.giveCard())
+    }
```
덱에서 카드를 뽑는다는 것을 코드를 들어가서 보지 않으면 알기 어려워 보여요!
object인 Deck을 도메인 모델 여기저기서 사용한다면 어떤 어려움이 있을 수 있을까요?

medAndro replied:
의도치 않은 도메인에서 접근해서 덱이 보유한 카드를 소진시켜 버리는 문제가 발생하고
Deck이 여러 도메인과 결합되어 추후에 수정하기 까다로워질 같네요
그리고 항상 Deck이 랜덤으로만 카드를 보관하고 있어서 테스트 하기 어려운것 같습니다.

Deck이 Object가 아니라고 한다면 `Deck.giveCard()` 같은 코드는 애초에 불가능하고 
컨트롤러에서 아래와 같이 덱 인스턴스에서 뽑아 직접 꽂아주며
```kt
player1.handCards.addCard(deck.giveCard())
```
GameParticipant 입장에서는 덱이 주든 딜러가주든 관심 없게끔 해야 할 것 같네요

hyemdooly replied:
handCards에 다이렉트로 꽂기보다, drawCard에 넘기는 형태면 더 객체가 일을 하는 형태가 될 것 같아요!

hyemdooly commented on src/main/kotlin/dslstudy/LanguagesBuilder.kt:
DSL Study로 사용한 코드들은 test 코드에 같이 포함해도 되지 않을까요?
test에만 사용하고 있으니까요!

medAndro replied:
듣고보니 test로 옮겨야 할 것 같네요... 올기도록 하겠습니다 🫡

hyemdooly commented on src/test/kotlin/blackjack/domain/model/card/DeckTest.kt:
```diff
@@ -0,0 +1,13 @@
+
+class DeckTest {
+    @Test
+    fun `카드를 뽑을 수 있다`() {
```
`giveCard`를 했는데 안에서 카드를 생성해서 주기만 한다면 어떨까요?
유효한 테스트 코드라고 할 수 있을까요?

medAndro replied:
`덱의 카드를 한장 줄 수 있다`라고 바꾸고 카드의 수량이 줄어드는지도 검사해야 할 것 같습니다

hyemdooly commented on src/test/kotlin/blackjack/domain/model/participant/PlayerTest.kt:
```diff
@@ -0,0 +1,15 @@
+import org.junit.jupiter.api.Test
+
+class PlayerTest {
+    @Test
+    fun `플레이어는 덱에서 카드를 뽑을 수 있다`() {
+        val player = Player()
+
+        player.drawCard()
+
+        assertThat(player.cardSize()).isEqualTo(1)
```
이 코드만 보더라도 덱에서 카드를 뽑는지 겉으로 봐서는 알 수 없어요.
`drawCard` 함수에서 카드를 생성해서 추가하기만 해도, 이 테스트는 통과할거에요!

medAndro replied:
이것도 덱이 Object라서 발생하는 문제 같습니다
카드 인스턴스를 한장 전달받아 손패에 추가하면 애초에 이런 테스트가 안 생기겠네요

hyemdooly commented on src/main/kotlin/blackjack/controller/Casino.kt:
```diff
@@ -0,0 +1,82 @@
+        participants.forEach { participant ->
+            repeat(INIT_CARD_SIZE) { participant.drawCard() }
+        }
+        outputView.lineSeparator()
```
아주 사소한 의견인데요!
이런 함수는 다음에 나올 outputView 함수에 껴넣을 수 있을 것 같다는 생각이 들어요. 🤔 
콘솔이 아닌 다른 뷰로 플랫폼이 바뀐다면?! 이라는 생각이 드네요!

medAndro replied:
`outputView.lineSeparator()` 한정으로는 OutputView 책임인데 컨트롤러로 잘못 들어간게 맞다고 생각합니다
```kt
participants.forEach { participant ->
            repeat(INIT_CARD_SIZE) { participant.drawCard() }
        }
 ```
 까지 선택되어있는데 이게 outputView의 책임인지는 잘 모르겠네요...🤔

hyemdooly replied:
오우 제가 드래그를 잘못했군요?!
`blackjackGame()` 함수 안에서 호출하는 `outputView.showDistributeCardMessage()` 함수에 껴넣을 수 있을까?라는 의견이었습니다!

hyemdooly commented on src/main/kotlin/blackjack/view/Translator.kt:
이렇게 Locale에 따라 변환하는건 요구사항이 아니었는데 꼭 필요할까요? 🤔

medAndro replied:
`가령 2하트 라고출력하는 것을 화면에 따라 ❤️2 이렇게 출력할지도 모릅니다.`
라는 다른 리뷰어의 리뷰를 우연히 보고 Suit enum을 Output에서 그대로 받아 처리해야 한다고 생각했고
더 나아가 현재는 한국어로만 출력하지만 다르게 출력하려면 어떻게 해야 할까 고민해 보다가 저렇게 구현해 보았습니다 
이 방식도 현재는 그냥 Locale 상수를 가져다 썼는데 ❤️같은 아이콘으로 하려면 애매해 지겠네요
결론은 꼭 필요하지는 않지만 이미 구현해놓은걸 수정할 이유도 없어보입니다.

hyemdooly commented on src/main/kotlin/blackjack/controller/Casino.kt:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    companion object {
+        const val INIT_CARD_SIZE = 2
```
초기 카드 장수를 컨트롤러가 들고 있는 이유는 무엇일까요?

medAndro replied:
플레이어의 카드를 초기화하는 등의 게임 진행 역할은 컨트롤러가 맡아야 하고 그때 필요한 상수라서 들고 있게 했던 것 같습니다. 
초기 카드가 2장이라는 것은 게임의 본질적인 규칙이라 도메인이 가지는게 맞을 것 같고 HandCards에게 줄 것 같습니다.

hyemdooly commented on src/main/kotlin/blackjack/controller/Casino.kt:
```diff
@@ -0,0 +1,82 @@
+        players.forEach { player ->
+            while (player.handCards.getStatus() != CardStatus.BUST && isPlayerWantHit(player)) {
+                player.drawCard()
+                outputView.showPlayerCardsInfo(player)
+            }
+            if (player.cardSize() == INIT_CARD_SIZE) {
+                outputView.showPlayerCardsInfo(player)
```
이 OutputView 함수를 한번만 쓰도록 수정해볼 수 있을까요?

medAndro replied:
몇시간 고민해봤는데 답을 못찾겠습니다.

히트를 하면 한장을 드로우하고 카드 정보를 표시해야하고
히트를 한번 이상 하고 스테이 할 경우 카드 정보를 표시하지 말하야 하고 
히트를 한번도 하지 않고 스테이 했을떄는 카드 정보를 표시해야 하는걸 다 지켜가면서 
`outputView.showPlayerCardsInfo(player)`를 한번만 쓰도록 수정하기가 까다롭네요... 🤔

```kt
private fun runPlayersDrawPhase(players: List<Player>) {
    players.forEach { player ->
        while (player.handCards.getStatus() != CardStatus.BUST) {
            val isHit = isPlayerWantHit(player)
            if (isHit) {
                player.drawCard()
            }
            if (isHit || player.cardSize() == INIT_CARD_SIZE) {
                outputView.showPlayerCardsInfo(player)
            }
            if (!isHit) {
                break
            }
        }
    }
    outputView.lineSeparator()
}
```

이런 코드는 짰는데 depth문제도 있고 딱봐도 정답은 아닌것 같네요...
힌트 부탁드립니다.

hyemdooly replied:
> 히트를 한번 이상 하고 스테이 할 경우 카드 정보를 표시하지 말하야 하고

아 제가 이 조건을 놓쳤군요 😢😢😢 저도 해봤는데 메다가 올려주신 코드가 최선이네요..
이 리뷰는 넘겨주세요! 정말정말 미안해요!!🧎‍♀️

medAndro replied:
아니에요 고민하는 과정에서 저도 공부가 많이 되었어요 ☺️

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/HandCards.kt:
```diff
@@ -0,0 +1,37 @@
+import blackjack.domain.model.card.Number.ACE
+import blackjack.domain.model.participant.CardStatus
+
+class HandCards {
```
손에 쥐고 있는 카드들이 직접 일을 하고 있네요 👍

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/HandCards.kt:
```diff
@@ -0,0 +1,37 @@
+class HandCards {
+    private val cards: MutableList<Card> = mutableListOf()
+
+    fun currentCards(): List<Card> = cards.toList()
```
backing property에 대해 배우셨다면 활용해보세요!

medAndro replied:
반영해 보겠습니다 🫡

hyemdooly commented on src/test/kotlin/blackjack/domain/model/card/HandCardsTest.kt:
```diff
@@ -0,0 +1,48 @@
+    fun `특정 인덱스의 카드를 반환할 수 있다`() {
+        val handCards = HandCards()
+
+        handCards.addCard(Card(Number.ACE))
+        handCards.addCard(Card(Number.QUEEN))
+        handCards.addCard(Card(Number.SEVEN))
```
초기값을 세팅하기 위해 일일히 add하고 있군요!
더 간단하게 세팅할 방법이 없을까요?

medAndro replied:
```kt
class HandCards(
    private val cards: MutableList<Card> = mutableListOf(),
)
```
로 HandCards를 수정하면 간단해 질 것 같아요...!

hyemdooly commented on src/main/kotlin/blackjack/domain/model/participant/CardStatus.kt:
```diff
@@ -0,0 +1,22 @@
+    companion object {
+        const val BLACKJACK_NUMBER = 21
+
+        fun calculateCardsStatus(cards: Collection<Card>): CardStatus =
+            when {
+                cards.size == INIT_CARD_SIZE && cards.sumOf { it.getMaximumValue() } == BLACKJACK_NUMBER -> BLACKJACK
+                cards.sumOf { it.getMinimumValue() } > BLACKJACK_NUMBER -> BUST
+                else -> NORMAL
+            }
+    }
```
손에 쥔 카드들이 직접 자신의 상태를 계산한다면 어떨까요?

medAndro replied:
`CardStatus`가 직접 카드를 보고 `CardStatus`를 알려주는게 직관적인거 같고 이곳 저곳에서 불러다가 사용할 것 같아서 추가했었습니다. 
`calculateCardsStatus()`는 `HandCards` 클래스 에서만 사용하고 있고 앞으로도 그럴 것 같아서 옮겨볼게요!

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/Number.kt:
```diff
@@ -0,0 +1,33 @@
+package blackjack.domain.model.card
+
+enum class Number(
+    val value: List<Int>,
+    val orderNumber: Int,
+    val initial: String,
+) {
+    ACE(listOf(1, 11), 1, "A"),
+    TWO(listOf(2), 2, "2"),
+    THREE(listOf(3), 3, "3"),
+    FOUR(listOf(4), 4, "4"),
+    FIVE(listOf(5), 5, "5"),
+    SIX(listOf(6), 6, "6"),
+    SEVEN(listOf(7), 7, "7"),
+    EIGHT(listOf(8), 8, "8"),
+    NINE(listOf(9), 9, "9"),
+    TEN(listOf(10), 10, "10"),
+    JACK(listOf(10), 11, "J"),
+    QUEEN(listOf(10), 12, "Q"),
+    KING(listOf(10), 13, "K"),
+    ;
```
`orderNumber`을 선언해주신 이유가 있으실까요?
순서때문이라면 `enum class`가 제공해주는 것들을 사용해서 다른 방법이 있을 것 같아요!
(다른 리뷰에서 계속...)

medAndro replied:
[첫 PR에서의 CardNumber 코드](https://github.com/medAndro/kotlin-blackjack/blob/4249cd80fd8bfd1bac873e3bcafa3b0db37c5dc1/src/main/kotlin/blackjack/domain/model/CardNumber.kt) 의 cardNumber가 실제 카드번호와 같이 1부터 시작하게 했어서 이것을 유지시키며 리팩터링 하다보니 1부터 시작하는 값을 가지고 있는것이 좋겠다고 생각하여 추가했었는데
```kt
fun getByOrderNumber(orderNumber: Int): Number = entries[orderNumber - 1] // ( +에서 -로 수정)
```
와 같이 [getByOrderNumber](https://github.com/medAndro/kotlin-blackjack/blob/f56ee63a3258ec3843af1e9b1d4aea049f6b6d03/src/main/kotlin/blackjack/domain/model/card/Number.kt#L24) 메서드를 수정하고 `orderNumber`는 삭제해야 겠습니다!

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/Card.kt:
```diff
@@ -0,0 +1,27 @@
+    val number: Number,
+    val suit: Suit = Suit.SPADE,
+) {
+    constructor(cardIndex: Int) : this(
+        number = Number.getByOrderNumber((cardIndex) % MAX_ORDER_NUMBER + 1),
+        suit = Suit.getBySuitIndex((cardIndex) / MAX_ORDER_NUMBER),
+    ) {
+        require(cardIndex in CARD_INDEX_RANGE) { ERROR_OUT_OF_CARD_INDEX }
+    }
```
Card를 생성할 때, cardIndex를 받아서 이렇게 계산해야만 할까요? 🤔
사실 카드 몇 번째가 어떤 문양과 번호인지는 관심사가 아닌 것 같은데.. 복잡한 것 같아요…

medAndro replied:
Deck에서 
```kt
private val cards = ArrayDeque(Card.CARD_INDEX_RANGE.shuffled().map { Card(it) })
```
와 같이 섞여있는 카드를 쉽게 만들 수 있게 하기 위해서 고안된 부 생성자입니다

전체 카드 인스턴스를 순서대로 일단 생성하고 섞는 방법도 생각해 보았는데
카드번호와 문양 조합 전부를 한장씩 가진 카드 컬렉션을 이중 for문을 사용하거나 flatMap을 사용해하여 생성한 뒤 섞는 방법만 생각나서 오히려 더 복잡해질 것 같아 이렇게 처리했습니다. 

더 쉬운 방법이 떠오르진 않네요...

hyemdooly commented on src/main/kotlin/blackjack/domain/model/card/Deck.kt:
```diff
@@ -0,0 +1,12 @@
+package blackjack.domain.model.card
+
+object Deck {
+    private val cards = ArrayDeque(Card.CARD_INDEX_RANGE.shuffled().map { Card(it) })
```
Deck 안에서 카드를 생성하는 것이 자연스러운 걸까요? 🤔

medAndro replied:
블랙잭 모듈 내에서 싱글턴으로 랜덤한 카드를 들고 있어야 하는 환경에서는 자연스러운 것 같은데
외부에서 카드 컬렉션을 받아서 초기화 되는 덱이라면 이상한 것 같네요...

hyemdooly reviewed (approved):
안녕하세요, 메다! 1단계 미션 고생하셨습니다. 😄 
전반적으로 좀 더 객체들의 티키타카를 고민할 수 있다면 좋을 것 같아요!
2단계 진행하시면서 함께 고민해보셨으면 하는 내용들을 리뷰로 드렸습니다!
궁금한 점은 언제든지 DM 또는 코멘트 달아주세요. 2단계도 화이팅입니다. 💪
